native/packages/react-native-bpk-component-animate-height/src/BpkAnimateHeight.js

```diff
@@ -28,114 +28,118 @@ class BpkAnimateHeight extends React.Component {
     super(props);
 
     this.state = {
-      expanded: this.props.expanded,
       height: new Animated.Value(COLLAPSED_HEIGHT),
-      inTree: this.props.expanded,
-      expandAnimationInProgress: false,
     };
 
+    this.inTree = this.props.expanded;
+    this.expandAnimationInProgress = false;
+
     // overflow: hidden is not properly supported on Android,
     // which causes animation to look shoddy.
     // See https://facebook.github.io/react-native/releases/0.49/docs/layout-props.html#overflow
     this.animationDuration =
       Platform.OS === 'ios' ? props.animationDuration : 0;
-
-    this.componentWillReceiveProps = this.componentWillReceiveProps.bind(this);
-    this.setExpandedHeight = this.setExpandedHeight.bind(this);
-    this.onHeightAnimationComplete = this.onHeightAnimationComplete.bind(this);
-    this.performAnimation = this.performAnimation.bind(this);
   }
 
   componentWillReceiveProps(nextProps) {
-    if (nextProps.expanded !== this.state.expanded) {
-      if (this.state.expanded) {
-        this.setState({ expanded: false });
-        this.performAnimation(COLLAPSED_HEIGHT, this.props.collapseDelay);
-      } else {
-        this.setState({ inTree: true, expandAnimationInProgress: true });
-      }
+    if (nextProps.expanded === this.props.expanded) {
+      return;
+    }
+
+    if (nextProps.expanded) {
+      this.inTree = true;
+      this.expandAnimationInProgress = true;
+    } else {
+      this.animate(COLLAPSED_HEIGHT, this.props.collapseDelay);
     }
   }
 
-  onHeightAnimationComplete() {
+  onHeightAnimationComplete = () => {
     if (typeof this.props.onAnimationComplete === 'function') {
       this.props.onAnimationComplete();
     }
+
     // eslint-disable-next-line no-underscore-dangle
     if (this.state.height._value - 0.01 < 0.00001) {
-      this.setState({ inTree: false });
+      this.inTree = false;
+      this.forceUpdate();
     }
```

Comment:
I'm wondering if we can avoid using `forceUpdate` as React [discourages it](https://reactjs.org/docs/react-component.html#forceupdate). Is `this.setState(this.state)` too hacky?

Reply:
The React docs indeed discourage usage of `forceUpdate`, but putting things in state that don't belong in state is also discouraged. In this particular example I am using `forceUpdate` to trigger a `render` in order to remove the view from the react tree. Previously this was handled by `setState({ inTree: false });` But `inTree` doesn't really belong in state. 

i.e. previously we'd call `setState({ inTree: true });` in `componentWillReceiveProps` - this actually results in 2 renders instead of one.

Reply:
oh and i guess to answer your specific question - `this.setState(this.state)` is no better than `this.forceUpdate()` lol